File: uservice-reports/main/src/main/scala/sinnet/report2/DtoDomainMapper.scala
package sinnet.report2

import sinnet.report2.grpc.{
  ReportRequest => ReportRequestDTO,
  ActivityDetails => ActivityDetailsDto}
import java.time.LocalDate
import scala.collection.JavaConverters._
import scala.util.Try
import sinnet.reports._

/** Converts ReportRequest (DTO) to its local model. */
object DtoDomainMapper {

  implicit def toActivityDetails(x: ActivityDetailsDto) = ActivityDetails(x.getYearMonth(), x.getPersonName(), Kilometers(x.getHowFarInKms()), Minutes(x.getHowLongInMins()));
    
  implicit def apply(dto: ReportRequestDTO): ReportRequest = {
    ReportRequest(dto.getDetailsList.asScala.map(toActivityDetails _))
  }
}

File: uservice-timeentries/host/src/main/java/sinnet/grpc/reports/report2/DtoDomainMapper.java
package sinnet.grpc.reports.report2;

import lombok.experimental.UtilityClass;
import sinnet.grpc.common.Mapper;
import sinnet.grpc.reports.report2.Models.ActivityDetails;
import sinnet.grpc.reports.report2.Models.ReportRequest;
import sinnet.models.ActionDuration;
import sinnet.models.Distance;

/** Converts ReportRequest (DTO) to its local model. */
@UtilityClass
class DtoDomainMapper {

  static ActivityDetails fromDto(sinnet.report2.grpc.ActivityDetails dto) {
    return new ActivityDetails(Mapper.fromDto(dto.getYearMonth()),
      dto.getPersonName(),
      Distance.of(dto.getHowFarInKms()),
      ActionDuration.of(dto.getHowLongInMins()));
  }
    
  static ReportRequest fromDto(sinnet.report2.grpc.ReportRequest dto) {
    return new ReportRequest(dto.getDetailsList().stream().map(it -> fromDto(it)).toList());
  }

}
